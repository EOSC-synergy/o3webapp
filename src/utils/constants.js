import {
    mean as calculateMean,
    std as calculateStd,
    median as calculateMedian,
    quantile as calculatePercentile, // TODO import actual percentile
} from "../services/math/math";

// index.js

/**
 * Stores the default model group.
 *
 * Modify this object to change the name or the models that appear when
 * the WebApp is initially loaded.
 * @constant {object}
 * @category Utils
 * @subcategory Constants
 */
export const DEFAULT_MODEL_GROUP = {
    groupId: null, // no valid id => add new group 
    groupName: "Example Group",
    modelList: [
        "CCMI-1_ACCESS_ACCESS-CCM-refC2",
        "CCMI-1_ACCESS_ACCESS-CCM-senC2fGHG",
        "CCMI-1_CCCma_CMAM-refC2"
    ]
};

// client.js
/** Stores the error message that is shown if no month is selected.
 * @constant {string}
 * @category Utils
 * @subcategory Constants
*/
export const NO_MONTH_SELECTED = "Invalid Selection: Selecting at least one month should make this error go away.";

// ReferenceSlice
export const DEFAULT_REF_MODEL = "SBUV_GSFC_merged-SAT-ozone";
export const DEFAULT_REF_YEAR = 1980;

// Section.js
/** Stores the name of the LatitudeBandSelector component as a Symbol.
 *  @constant {string}
 * @category Utils
 * @subcategory Constants
*/
export const LBS_Symbol = Symbol("LatitudeBandSelector");
/** Stores the name of the LocationSelector component as a Symbol.
 *  @constant {string}
 * @category Utils
 * @subcategory Constants
*/
export const LS_Symbol = Symbol("LocationSelector");
/** Stores the name of the ModelGroupConfigurator component as a Symbol.
 *  @constant {string}
 * @category Utils
 * @subcategory Constants
*/
export const MGC_Symbol = Symbol("ModelGroupConfigurator");
/** Stores the name of the OffsetConfigurator component as a Symbol.
 *  @constant {string}
 * @category Utils
 * @subcategory Constants
*/
export const OC_Symbol = Symbol("OffsetConfigurator");
/** Stores the name of the PlotNameField component as a Symbol.
 *  @constant {string}
 * @category Utils
 * @subcategory Constants
*/
export const PNF_Symbol = Symbol("PlotNameField");
/** Stores the name of the ReferenceModelSelector component as a Symbol.
 *  @constant {string}
 * @category Utils
 * @subcategory Constants
*/
export const RMS_Symbol = Symbol("ReferenceModelSelector");
/** Stores the name of the ReferenceYearField component as a Symbol.
 *  @constant {string}
 * @category Utils
 * @subcategory Constants
*/
export const RYF_Symbol = Symbol("ReferenceYearField");
/** Stores the name of the RegionSelector component as a Symbol.
 *  @constant {string}
 * @category Utils
 * @subcategory Constants
*/
export const RS_Symbol = Symbol("RegionSelector");
/** Stores the name of the TimeCheckBoxGroup component as a Symbol.
 *  @constant {string}
 * @category Utils
 * @subcategory Constants
*/
export const TCG_Symbol = Symbol("TimeCheckBoxGroup");
/** Stores the name of the XAxisField component as a Symbol.
 *  @constant {string}
 * @category Utils
 * @subcategory Constants
*/
export const XAF_Symbol = Symbol("XAxisField");
/** Stores the name of the YAxisField component as a Symbol.
 *  @constant {string}
 * @category Utils
 * @subcategory Constants
*/
export const YAF_Symbol = Symbol("YAxisField");
/** Stores the name of the CustomLatitudeSelector component as a Symbol.
 *  @constant {string}
 * @category Utils
 * @subcategory Constants
*/
export const CLS_Symbol = Symbol("CustomLatitudeSelector");

// TimeCheckboxGroup.js
/** Stores the season Winter and its corresponding months.
 *  @constant {object}
 * @memberof TimeCheckBoxGroup
*/
export const Winter = {name: Symbol("Winter"), months: [12, 1, 2], seasonId: 0}

/** Stores the season Spring and its corresponding months.
 * @constant {object}
 * @memberof TimeCheckBoxGroups
*/
export const Spring = {name: Symbol("Spring"), months: [3, 4, 5], seasonId: 1}

/** Stores the season Summer and its corresponding months.
 * @constant {object}
 * @memberof TimeCheckBoxGroup
*/
export const Summer = {name: Symbol("Summer"), months: [6, 7, 8], seasonId: 2}

/** Stores the season Autumn and its corresponding months.
 * @constant {object}
 * @memberof TimeCheckBoxGroup
*/
export const Autumn = {name: Symbol("Autumn"), months: [9, 10, 11], seasonId: 3}

/** Array containing all season indext with the corresponding seasonId
 * @constant {array}
 * @memberof TimeCheckBoxGroup
*/
export const SEASONS_ARRAY = [Winter, Spring, Summer, Autumn];

/** Stores all the months in short form. (first three letters)
 * @constant {array}
 * @category Utils
 * @subcategory Constants
*/
export const months = [
    Symbol("Jan"), Symbol("Feb"), Symbol("Mar"), Symbol("Apr"),
    Symbol("May"), Symbol("Jun"), Symbol("Jul"), Symbol("Aug"),
    Symbol("Sep"), Symbol("Oct"), Symbol("Nov"), Symbol("Dec")
]
/** Stores the amount of months in a season
 * @constant {number}
 * @category Utils
 * @subcategory Constants
*/
export const NUM_MONTHS_IN_SEASON = 3;

/** Stores the amount of months in a year
 * @constant {number}
 * @category Utils
 * @subcategory Constants
 */
export const NUM_MONTHS = 12;

/** These months are selected when the webapp starts
 * @constant {array}
 * @category Utils
 * @subcategory Constants
*/
export const DEFAULT_MONTHS = [1, 2, 12];

// LatitudeBandSelector.js
/** Stores the latitude bands and its min/max values
 * @constant {array}
 * @category Utils
 * @subcategory Constants
*/
export const latitudeBands = [
    {
        text: Symbol("Southern Hemisphere (SH) Polar (90–60°S)"),
        value: {minLat: -90, maxLat: -60}
    },
    {
        text: Symbol("SH Mid-Latitudes (60–35°S)"),
        value: {minLat: -60, maxLat: -35}
    },
    {
        text: Symbol("Tropics (20°S–20°N)"),
        value: {minLat: -20, maxLat: 20}
    },
    {
        text: Symbol("Northern Hemisphere (NH) Mid-Latitudes (35–60°N)"),
        value: {minLat: 35, maxLat: 60}
    },
    {
        text: Symbol("NH Polar (60–90°N)"),
        value: {minLat: 60, maxLat: 90}
    },
    {
        text: Symbol("Near-Global (60°S–60°N)"),
        value: {minLat: -60, maxLat: 60}
    },
    {
        text: Symbol("Global (90°S–90°N)"),
        value: {minLat: -90, maxLat: 90}
    },
    // !!! Custom must be last in array !!!
    {
        text: Symbol("Custom"),
        value: 'custom'
    },
]

export const LATITUDE_BAND_LIST = latitudeBands.map(obj => obj.text.description);

// DownloadModal.js
export const fileFormats = [
    Symbol("CSV"), Symbol("PDF"), Symbol("PNG"), Symbol("SVG")
];

// PlotNameField.js

/** The max. length of the plot name
 * @constant {number}
 * @category Utils
 * @subcategory Constants
*/
export const PLOT_NAME_MAX_LEN = 40;
/*
// GRAPH
*/

/**
 * An "enum" which stores the plot types as they come from the api.
 * @constant {object}
 * @category Utils
 * @subcategory Constants
 */
export const O3AS_PLOTS = { // used for internal testing or manual if-else
    tco3_zm: "tco3_zm",
    tco3_return: "tco3_return",
};

/**
 * Maps the plots provided by the api to their apexcharts plot type
 * @constant {object}
 * @category Utils
 * @subcategory Constants
 */
export const APEXCHART_PLOT_TYPE = {
    tco3_zm: "line",
    tco3_return: "boxPlot"
};

/**
 * How large the loading spinner should appear.
 * @constant {string}
 * @category Utils
 * @subcategory Constants
 */
export const HEIGHT_LOADING_SPINNER = "300px";
/**
 * How tall the graph should appear
 * @constant {string}
 * @category Utils
 * @subcategory Constants
 */
export const HEIGHT_GRAPH = `${window.innerHeight * 0.75}px`;

/*
// Options Formatter, XAxisField, YAxisField, apiSlice
*/
/**
 * The start year, no data is fetched for years before this point in time
 * @constant {number}
 * @category Utils
 * @subcategory Constants
 */
export const START_YEAR = 1960;
/**
 * The end year, no data is fetched for years after this point in time
 * @constant {number}
 * @category Utils
 * @subcategory Constants
 */
export const END_YEAR = 2100;
/**
 * This array provides a list from start to end year with each year as a string
 * @constant {array}
 * @category Utils
 * @subcategory Constants
 */
export const IMPLICIT_YEAR_LIST = [...Array(END_YEAR - START_YEAR + 1).keys()].map(number => `${START_YEAR + number}`);

// important for api data transformation
/** This string stores the description of the antarctic region
 * @constant {string}
 * @category Utils
 * @subcategory Constants
*/
const ANTARCTIC = "Antarctic(Oct)";
/** This string stores the description of the south hemisphere region
 * @constant {string}
 * @category Utils
 * @subcategory Constants
*/
const SH_MID = "SH mid-lat";
/** This string stores the description of the north hemisphere region
 * @constant {string}
 * @category Utils
 * @subcategory Constants
*/
const NH_MID = "NH mid-lat";
/** This string stores the description of the tropics region
 * @constant {string}
 * @category Utils
 * @subcategory Constants
*/
const TROPICS = "Tropics";
/** This string stores the description of the arctic region
 * @constant {string}
 * @category Utils
 * @subcategory Constants
*/
const ARCTIC = "Arctic(Mar)";
/** This string stores the description of the near global region
 * @constant {string}
 * @category Utils
 * @subcategory Constants
*/
const NEAR_GLOBAL = "Near global";
/** This string stores the description of the global region
 * @constant {string}
 * @category Utils
 * @subcategory Constants
*/
const GLOBAL = "Global";
/** This string stores the description of the user region
 * @constant {string}
 * @category Utils
 * @subcategory Constants
*/
export const USER_REGION = "User region";

export const O3AS_REGIONS = {
    ANTARCTIC,
    SH_MID,
    NH_MID,
    TROPICS,
    ARCTIC,
    NEAR_GLOBAL,
    GLOBAL,
    // user region must be last element
    USER_REGION
}

// user region must be last element
export const ALL_REGIONS_ORDERED = [ANTARCTIC, SH_MID, NH_MID, TROPICS, ARCTIC, NEAR_GLOBAL, GLOBAL, USER_REGION];

/**
 * The mean: this appears in the model group card and is used to identify its statistical value settings.
 * @constant {string}
 * @category Utils
 * @subcategory Constants
 */
export const mean = "mean";
/**
 * The standard deviation: this appears in the model group card and is used to identify its statistical value settings.
 * @constant {string}
 * @category Utils
 * @subcategory Constants
 */
export const std = "standard deviation";
/**
 * The median: this appears in the model group card and is used to identify its statistical value settings.
 * @category Utils
 * @subcategory Constants
 * @constant {string}
 */
export const median = "median";
/**
 * The percentile: this appears in the model group card and is used to identify its statistical value settings.
 * @constant {string}
 * @category Utils
 * @subcategory Constants
 */
export const percentile = "percentile";

/**
 * The statistical values that are computable are listed here as
 * an "enum"
 * @constant {object}
 * @category Utils
 * @subcategory Constants
 */
export const STATISTICAL_VALUES = {}
// populate object automatically with the values: 
// the order corresponds to the order in the interface
const asList = [mean, std, median, percentile];
asList.forEach(sv => STATISTICAL_VALUES[sv] = sv);

/**
 * Name for the helper statistical value series: This line is used to calculate the mean+/-std
 * @constant {string}
 * @category Utils
 * @subcategory Constants
 */
export const stdMean = "stdMean";
/**
 * The lower percentile ~15.87th
 * @constant {string}
 * @category Utils
 * @subcategory Constants
 */
export const lowerPercentile = "lowerPercentile";
/**
 * The lower percentile ~84.13th
 * @constant {string}
 * @category Utils
 * @subcategory Constants
 */
export const upperPercentile = "upperPercentile";
/*
 * Regarding the percentile (message from Tobias):
 *
 * It would be good if you could include the 84.13th and 15.87th percentiles
 * which corresponds +1σ standard dev of the Gaussian distribution.
 * A workflow to calculate the percentiles is described on the following
 * website for example:
 *
 * https://www.indeed.com/career-advice/career-development/how-to-calculate-percentile
*/

/**
 * This extended sv list is used in the calculation to handle some changes that
 * came up after the specification / design phase: standard deviation is not one
 * but TWO lines. Same goes for percentile which is lower- AND upper-percentile.
 * @constant {array}
 * @category Utils
 * @subcategory Constants
 */
export const EXTENDED_SV_LIST = [stdMean, lowerPercentile, upperPercentile];


/**
 * The same statistical values as a list to verify certain payload data.
 * @constant {array}
 * @category Utils
 * @subcategory Constants
 */
export const STATISTICAL_VALUES_LIST = Object.values(STATISTICAL_VALUES);


/**
 * This object maps each statistical value that should be calculated
 * to a corresponding function describing HOW it should be calculated.
 * @constant {object}
 * @category Utils
 * @subcategory Constants
 */
export const SV_CALCULATION = {
    mean: calculateMean,
    median: calculateMedian,
    percentile: calculatePercentile,
    stdMean: calculateMean, // mean for std+-
}
SV_CALCULATION[std] = calculateStd;
SV_CALCULATION[lowerPercentile] = arr => calculatePercentile(arr, .1587);
SV_CALCULATION[upperPercentile] = arr => calculatePercentile(arr, .8413);
SV_CALCULATION[stdMean] = calculateMean;

/**
 * This object maps each statistical value that should be calculated
 * to a color it should be appear in.
 * @constant {object}
 * @category Utils
 * @subcategory Constants
 */
export const SV_COLORING = {
    mean: "#696969",
    "standard deviation": "#0e4e78",//"#000",
    median: "#000",
    percentile: "#000",
    "lowerPercentile": "#1e8509",
    "upperPercentile": "#1e8509",
    "mean+std": "#0e4e78",
    "mean-std": "#0e4e78",
}

/**
 * This object maps each statistical value that should be calculated
 * to its line dashing allowing an easy customization if e.g. the
 * mean should be dashed too.
 *
 * The integer values correspond to the dashing format that is
 * expected by apexcharts.
 * @constant {object}
 * @category Utils
 * @subcategory Constants
 */
export const SV_DASHING = {
    mean: 0,
    median: 2,
    percentile: 0,
    "mean+std": 8,
    "mean-std": 8,
    "lowerPercentile": 4,
    "upperPercentile": 4,
}

/**
 * This object maps each statistical value to its corresponding name that 
 * should be displayed in the legend of the plot and inside the tooltip when
 * hovering over a datapoint.
 */
export const SV_DISPLAY_NAME = {
    mean: "Mean",
    median: "Median",
    percentile: "Percentile",
    "mean+std": "μ + σ",
    "mean-std": "μ - σ",
    "lowerPercentile": "Lower %",
    "upperPercentile": "Upper %",

}

/**
 * This parameter specifies how thick the line of plotted models should appear.
 * @constant {number}
 * @category Utils
 * @subcategory Constants
 */
export const MODEL_LINE_THICKNESS = 2;
/**
 * This parameter specifies how thick the line of plotted statistical values should appear.
 * @constant {number}
 * @category Utils
 * @subcategory Constants
 */
export const STATISTICAL_VALUE_LINE_THICKNESS = 2;

/**
 * the Legal Notice links which will be parsed into the PDF.
 * @constant {array}
 * @category Utils
 * @subcategory Constants
 */
export const legalNoticeLinks = [
    "Terms of Use Link: https://o3as.data.kit.edu/policies/terms-of-use.html",
    "Privacy Policy Link: https://o3as.data.kit.edu/policies/privacy-policy.html",
    "How to Acknowledge Link: https://o3as.data.kit.edu/policies/how-to-acknowledge.html",
];
// Custom Latitude Selector

/**
 * The smallest possible latitude value.
 * @constant {number}
 * @category Utils
 * @subcategory Constants
 */
export const LATITUDE_BAND_MAX_VALUE = 90;

/**
 * The biggest possible latitude value.
 * @constant {number}
 * @category Utils
 * @subcategory Constants
 */
export const LATITUDE_BAND_MIN_VALUE = -90;


/**´
 * The default color of the backgrounds (Navbar, Footer, Sidebar, ...)
 * @constant {string}
 */
export const BACKGROUND_BASE_COLOR = '#111';
