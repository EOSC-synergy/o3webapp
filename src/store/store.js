import { applyMiddleware, combineReducers, configureStore, createStore } from '@reduxjs/toolkit';
import plotReducer from './plotSlice/plotSlice';
import modelsReducer, { setModelsOfModelGroup } from './modelsSlice/modelsSlice';
import referenceReducer from './referenceSlice/referenceSlice';
import apiReducer from '../services/API/apiSlice/apiSlice';
import { DEFAULT_MODEL_GROUP } from '../utils/constants';
import { HYDRATE, createWrapper } from 'next-redux-wrapper';
import thunkMiddleware from 'redux-thunk';

const bindMiddleware = (middleware) => {
    if (process.env.NODE_ENV !== 'production') {
        const { composeWithDevTools } = require('redux-devtools-extension');
        return composeWithDevTools(applyMiddleware(...middleware));
    }
    return applyMiddleware(...middleware);
};

const combinedReducer = combineReducers({
    plot: plotReducer,
    models: modelsReducer,
    reference: referenceReducer,
    api: apiReducer,
});

const reducer = (state, action) => {
    if (action.type === HYDRATE) {
        const nextState = {
            ...state, // use previous state
            ...action.payload, // apply delta from hydration
        };
        // if (state.count.count) nextState.count.count = state.count.count; // preserve count value on client side navigation
        return nextState;
    } else {
        return combinedReducer(state, action);
    }
};

let store;

const initStore = () => {
    store = createStore(reducer, bindMiddleware([thunkMiddleware]));
    return store;
};

export { store };

export const wrapper = createWrapper(initStore);

export function createTestStore() {
    const store = configureStore({
        reducer: combinedReducer,
    });
    store.dispatch(setModelsOfModelGroup(DEFAULT_MODEL_GROUP)); // fill with default group
    return store;
}
