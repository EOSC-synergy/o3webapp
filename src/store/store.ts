import { Action, combineReducers, configureStore, ThunkAction } from '@reduxjs/toolkit';
import plotReducer, { GlobalPlotState } from './plotSlice';
import modelsReducer, { GlobalModelState, setModelsOfModelGroup } from './modelsSlice';
import referenceReducer, { GlobalReferenceState } from './referenceSlice';
import apiReducer from '../services/API/apiSlice/apiSlice';
import { DEFAULT_MODEL_GROUP } from '../utils/constants';
import { createWrapper, HYDRATE } from 'next-redux-wrapper';
import thunkMiddleware from 'redux-thunk';
import { useDispatch, useStore } from 'react-redux';

type GlobalState =
    | undefined
    | (GlobalPlotState & GlobalModelState & GlobalReferenceState & { api: any });

const combinedReducer = combineReducers({
    plot: plotReducer,
    models: modelsReducer,
    reference: referenceReducer,
    api: apiReducer,
});

const reducer = (state: GlobalState, action: { type: string; payload: any }) => {
    if (action.type === HYDRATE) {
        // if (state.count.count) nextState.count.count = state.count.count; // preserve count value on client side navigation
        return {
            ...state, // use previous state
            ...action.payload, // apply delta from hydration
        };
    } else {
        return combinedReducer(state, action);
    }
};

//let store: ReturnType<typeof configureStore<GlobalState, >>;

const initStore = () => {
    //store = createStore(reducer, bindMiddleware([thunkMiddleware]));
    return configureStore({
        reducer,
        devTools: process.env.NODE_ENV !== 'production',
        middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(thunkMiddleware),
    });
};

export type AppStore = ReturnType<typeof initStore>;
export type AppState = ReturnType<AppStore['getState']>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action>;

export type AppDispatch = AppStore['dispatch'];
export const useAppDispatch = () => useDispatch<AppDispatch>(); // Export a hook that can be reused to resolve types

export const useAppStore = () => useStore<AppState>();

export const wrapper = createWrapper(initStore);

export function createTestStore() {
    const store = initStore();
    store.dispatch(setModelsOfModelGroup(DEFAULT_MODEL_GROUP)); // fill with default group
    return store;
}
