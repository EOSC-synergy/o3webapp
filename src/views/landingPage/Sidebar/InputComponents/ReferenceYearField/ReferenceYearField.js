import React from "react";
import {Grid, Typography, FormControl, TextField, Checkbox} from "@mui/material";
import {useDispatch} from "react-redux"
import {useSelector} from "react-redux";
import {END_YEAR, START_YEAR, O3AS_PLOTS} from "../../../../../utils/constants";
import {fetchPlotDataForCurrentModels} from "../../../../../services/API/apiSlice";
import {setYear, setVisibility, selectRefYear, selectVisibility} from "../../../../../store/referenceSlice/referenceSlice";
import VisibilityOffIcon from '@mui/icons-material/VisibilityOff';
import MuiVisibilityIcon from '@mui/icons-material/Visibility';
import { selectPlotId } from "../../../../../store/plotSlice/plotSlice";

/**
 * Enables the user to select a reference year.
 * @component
 * @param {Object} props
 * @param {function} props.reportError - function to handle errors
 * @returns {JSX.Element} a jsx containing a text field to select the reference year
 */
function ReferenceYearField(props) {
    /**
     * A dispatch function to dispatch actions to the redux store.
     */
    const dispatch = useDispatch();

    /**
     * The selected reference year from the redux store.
     */
    const selectedYear = useSelector(selectRefYear);
    const refLineVisibility = useSelector(selectVisibility);
    const plotId = useSelector(selectPlotId);

    /**
     * Handles the change of the reference year field if it is modified.
     */
    const handleChangeForRefYear = (event) => {
        if (!isNaN(event.target.value)) {
            dispatch(setYear({year: event.target.value}));
            if (event.target.value >= START_YEAR && event.target.value <= END_YEAR) {
                // fetch for tco3_zm and tco3_return
                dispatch(fetchPlotDataForCurrentModels());
            }
        }
    };

    /**
     * Handles the change of the reference line visibility field if it is modified.
     */
    const handleShowRefLineClicked = (event) => {
        dispatch(setVisibility({visible: event.target.checked}))
    }

    return (
        <>
            <Grid container sx={{width: "90%", marginLeft: "auto", marginRight: "auto", marginTop: "5%"}}>
                <Grid item xs={5}>
                    <Typography>Reference Year:</Typography>
                </Grid>
                <Grid item xs={7} sx={{mt: "-8px"}}>
                    <FormControl sx={{width: '35%'}}>
                        <TextField
                            variant="outlined"
                            id="outlined-basic"
                            size="small"
                            value={selectedYear}
                            onChange={handleChangeForRefYear}
                            error={selectedYear < START_YEAR || selectedYear > END_YEAR}
                            helperText={selectedYear < START_YEAR ? `<${START_YEAR}` : (selectedYear > END_YEAR ? `>${END_YEAR}` : '')}
                        />
                    </FormControl>
                    {
                        plotId === O3AS_PLOTS.tco3_zm
                        && 
                        <FormControl>
                            <Checkbox
                                checked={refLineVisibility}
                                icon={<VisibilityOffIcon data-testid="RefLineInvisibleCheckbox"/>}
                                checkedIcon={<MuiVisibilityIcon/>}
                                onClick={handleShowRefLineClicked}
                            />
                        </FormControl>
                    }   
                </Grid>
            </Grid>
        </>
    );
}

export default ReferenceYearField;
