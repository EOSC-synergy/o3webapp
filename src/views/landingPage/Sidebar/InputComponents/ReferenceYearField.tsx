import React, { ChangeEvent } from 'react';
import { Grid, Typography, FormControl, TextField, Checkbox } from '@mui/material';
import { useSelector } from 'react-redux';
import { END_YEAR, START_YEAR, O3AS_PLOTS } from 'utils/constants';
import { fetchPlotDataForCurrentModels } from 'services/API/apiSlice/apiSlice';
import { setYear, setVisibility, selectRefYear, selectVisibility } from 'store/referenceSlice';
import VisibilityOffIcon from '@mui/icons-material/VisibilityOff';
import MuiVisibilityIcon from '@mui/icons-material/Visibility';
import { selectPlotId } from 'store/plotSlice';
import { useAppDispatch } from 'store/store';

/**
 * Enables the user to select a reference year.
 * @component
 * @returns {JSX.Element} a jsx containing a text field to select the reference year
 */
const ReferenceYearField: React.FC = () => {
    /**
     * A dispatch function to dispatch actions to the redux store.
     * @constant {function}
     */
    const dispatch = useAppDispatch();

    /**
     * The selected reference year from the redux store.
     * @see {@link selectRefYear}
     * @constant {int}
     */
    const selectedYear = useSelector(selectRefYear);

    /**
     * Whether the reference line is visble. Taken from the redux store.
     * @constant {int}
     * @see {@link selectVisibility}
     */
    const refLineVisibility = useSelector(selectVisibility);
    /**
     * The id of the current plot. Taken from the redux store.
     * @constant {String}
     * @see {@link selectPlotId}
     */
    const plotId = useSelector(selectPlotId);

    /**
     * Handles the change of the reference year field if it is modified.
     * @function
     * @see {@link fetchPlotDataForCurrentModels}
     * @param {Event} event the event that triggered the call of this function.
     */
    const handleChangeForRefYear = (event: ChangeEvent<HTMLInputElement>) => {
        if (!isNaN(event.target.valueAsNumber)) {
            dispatch(setYear({ year: event.target.valueAsNumber }));
            if (
                event.target.valueAsNumber >= START_YEAR &&
                event.target.valueAsNumber <= END_YEAR
            ) {
                // fetch for tco3_zm and tco3_return
                dispatch(fetchPlotDataForCurrentModels());
            }
        }
    };

    /**
     * Handles the change of the reference line visibility field if it is modified.
     * @function
     * @param {Event} event the event that triggered the call of this function
     */
    const handleShowRefLineClicked = (event: ChangeEvent<HTMLInputElement>) => {
        dispatch(setVisibility({ visible: event.target.checked }));
    };

    return (
        <>
            <Grid
                container
                sx={{ width: '90%', marginLeft: 'auto', marginRight: 'auto', marginTop: '5%' }}
            >
                <Grid item xs={5}>
                    <Typography>Reference Year:</Typography>
                </Grid>
                <Grid item xs={7} sx={{ mt: '-8px' }}>
                    <FormControl sx={{ width: '35%' }}>
                        <TextField
                            variant="outlined"
                            id="outlined-basic"
                            size="small"
                            value={selectedYear}
                            onChange={handleChangeForRefYear}
                            error={selectedYear < START_YEAR || selectedYear > END_YEAR}
                            helperText={
                                selectedYear < START_YEAR
                                    ? `<${START_YEAR}`
                                    : selectedYear > END_YEAR
                                    ? `>${END_YEAR}`
                                    : ''
                            }
                            inputProps={{ 'data-testid': 'ReferenceYearField-year' }}
                            type="number"
                        />
                    </FormControl>
                    {plotId === O3AS_PLOTS.tco3_zm && (
                        <FormControl>
                            <Checkbox
                                checked={refLineVisibility}
                                icon={<VisibilityOffIcon data-testid="RefLineInvisibleCheckbox" />}
                                checkedIcon={<MuiVisibilityIcon />}
                                onChange={handleShowRefLineClicked}
                                inputProps={{
                                    // @ts-expect-error https://github.com/mui/material-ui/issues/20160
                                    'data-testid': 'ReferenceYearField-toggleVisibility',
                                }}
                            />
                        </FormControl>
                    )}
                </Grid>
            </Grid>
        </>
    );
};

export default ReferenceYearField;
