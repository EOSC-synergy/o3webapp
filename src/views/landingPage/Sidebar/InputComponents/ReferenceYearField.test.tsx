import React from 'react';
import ReferenceYearField from './ReferenceYearField';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { AppStore, createTestStore } from 'store/store';
import { Provider } from 'react-redux';
import { END_YEAR, START_YEAR } from 'utils/constants';
import userEvent from '@testing-library/user-event';

let store: AppStore;

beforeEach(() => {
    store = createTestStore();
});

describe('tests basic rendering', () => {
    it('Component renders without crashing', () => {
        render(
            <Provider store={store}>
                <ReferenceYearField />
            </Provider>
        );
    });

    // Snapshot test
    it('renders correctly', () => {
        const { container } = render(
            <Provider store={store}>
                <ReferenceYearField />
            </Provider>
        );
        expect(container).toMatchSnapshot();
    });
});

describe('tests redux functionality', () => {
    it('updates the reference year value in the store correctly', () => {
        const startValue = store.getState().reference.settings.year;
        const newValue = startValue + 20;

        const { getByTestId } = render(
            <Provider store={store}>
                <ReferenceYearField />
            </Provider>
        );
        const inputField = getByTestId('ReferenceYearField-year');
        fireEvent.change(inputField, { target: { value: newValue } }); // change input
        expect(store.getState().reference.settings.year).toEqual(newValue);
    });

    it('toggles the visibility of the reference line in the store correctly', () => {
        const startValue = store.getState().reference.settings.visible;

        const { getByTestId } = render(
            <Provider store={store}>
                <ReferenceYearField />
            </Provider>
        );
        const toggleButton = getByTestId('ReferenceYearField-toggleVisibility');
        userEvent.click(toggleButton); // toggle visibility
        expect(store.getState().reference.settings.visible).toEqual(!startValue);
    });
});

describe('tests error handling', () => {
    it('displays a warning if the reference year value is outside of the allowed range', () => {
        const { getByTestId, container } = render(
            <Provider store={store}>
                <ReferenceYearField />
            </Provider>
        );
        const inputField = getByTestId('ReferenceYearField-year');
        fireEvent.change(inputField, { target: { value: END_YEAR + 1 } }); // change input (not valid)
        expect(container).toHaveTextContent(`>${END_YEAR}`);
        fireEvent.change(inputField, { target: { value: START_YEAR - 1 } }); // change input (not valid)
        expect(container).toHaveTextContent(`<${START_YEAR}`);
    });
});
