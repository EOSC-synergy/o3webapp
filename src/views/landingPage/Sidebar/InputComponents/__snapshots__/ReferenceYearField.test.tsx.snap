// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests basic rendering renders correctly 1`] = `
<div>
  <div
    class="MuiGrid-root MuiGrid-container css-wts08w-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 css-k3f0g4-MuiGrid-root"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        Reference Year:
      </p>
    </div>
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-7 css-a0yaxr-MuiGrid-root"
    >
      <div
        class="MuiFormControl-root css-1st3y2y-MuiFormControl-root"
      >
        <div
          class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
        >
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
              data-testid="ReferenceYearField-year"
              id="outlined-basic"
              type="number"
              value="1980"
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-hdw1oc"
              >
                <span
                  class="notranslate"
                >
                  â€‹
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
      >
        <span
          class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root Mui-checked css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
        >
          <input
            checked=""
            class="PrivateSwitchBase-input css-1m9pwf3"
            data-indeterminate="false"
            data-testid="ReferenceYearField-toggleVisibility"
            type="checkbox"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="VisibilityIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </span>
      </div>
    </div>
  </div>
</div>
`;
